cpp/src/arrow/compute/exec/exec_plan.cc

```diff
@@ -46,7 +46,9 @@ namespace {
 struct ExecPlanImpl : public ExecPlan {
   explicit ExecPlanImpl(ExecContext* exec_context,
                         std::shared_ptr<const KeyValueMetadata> metadata = NULLPTR)
-      : ExecPlan(exec_context), metadata_(std::move(metadata)) {}
+      : ExecPlan(exec_context), metadata_(std::move(metadata)) {
+        span_ = arrow::internal::tracing::OTSpan();
```

Comment:
I'm not sure why, but we segfault in the first line of `StartProducing()`. I believe this is because I am not initialized `span_` correctly. In the header file, it's type is is the public Span class `arrow::util::tracing::Span`, but for the macros to work it needs to be initialized as an `arrow::internal::tracing::OTSpan` so that is has the shared_ptr field. (When the spans are used just internally, I was able to solve this same segfault by switching the declaration to `OTSpan`, but that won't work for the member variables.)

Any idea what might be going wrong?

Reply:
How about something like this: https://github.com/westonpace/arrow/commit/85090cc7f8f3da02c1bccac124989b5243c89d97

Then we can still use span but the details are hidden.

Reply:
Thanks, Weston. That works much better.